### PR TITLE
support dynamic shape for conv3d

### LINKED ISSUE #16834: Tensorflow not supporting 3D convolution with None input size
Hello,

I am trying to build a 3D fully connected convolutional network with input of variable size. To do that I define x with None and dimension and then apply a conv3d layers:

x = T.placeholder(shape=(1,nbrChannel,None,None,None),name="input")
conv=T.layers.conv3d(inputs=x,filters=num_filters,kernel_size=filter_size,strides=strides,padding="same",data_format='channels_first')

I get the following error. Please note that I tried to do exactly the same thing with 2D image input (with 2 "None" dimensions) and 2D convolution and it worked fine. Therefore, I think this is a problem specific to the conv3d layer function.

> 
> c:\programdata\miniconda3\lib\site-packages\tensorflow\python\layers\convolutional.py in conv3d(inputs, filters, kernel_size, strides, padding, data_format, dilation_rate, activation, use_bias, kernel_initializer, bias_initializer, kernel_regularizer, bias_regularizer, activity_regularizer, kernel_constraint, bias_constraint, trainable, name, reuse)
>     817       _reuse=reuse,
>     818       _scope=name)
> --> 819   return layer.apply(inputs)
>     820 
>     821 
> 
> c:\programdata\miniconda3\lib\site-packages\tensorflow\python\layers\base.py in apply(self, inputs, *args, **kwargs)
>     760       Output tensor(s).
>     761     """
> --> 762     return self.__call__(inputs, *args, **kwargs)
>     763 
>     764   def _add_inbound_node(self,
> 
> c:\programdata\miniconda3\lib\site-packages\tensorflow\python\layers\base.py in __call__(self, inputs, *args, **kwargs)
>     650 
>     651         if not in_deferred_mode:
> --> 652           outputs = self.call(inputs, *args, **kwargs)
>     653           if outputs is None:
>     654             raise ValueError('A layer\'s `call` method should return a Tensor '
> 
> c:\programdata\miniconda3\lib\site-packages\tensorflow\python\layers\convolutional.py in call(self, inputs)
>     182           outputs_4d = array_ops.reshape(outputs,
>     183                                          [outputs_shape[0], outputs_shape[1],
> --> 184                                           outputs_shape[2] * outputs_shape[3],
>     185                                           outputs_shape[4]])
>     186           outputs_4d = nn.bias_add(outputs_4d, self.bias, data_format='NCHW')
> 
> TypeError: unsupported operand type(s) for *: 'NoneType' and 'NoneType'

### LINKED ISSUE #20527: tf.layers.conv3d with "channels_first" does not accept dynamic shapes
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
v1.8.0-0-g93bc2e2072
- **Python version**: 
3.6
- **Bazel version (if compiling from source)**:
N/A
- **GCC/Compiler version (if compiling from source)**:
N/A
- **CUDA/cuDNN version**:
cuDNN 7.0 CUDA 9.1
- **GPU model and memory**:
Geforce 1080 Ti 12Gb
- **Exact command to reproduce**:
See given code snippets

### Describe the problem
Calling tf.layers.conv3d with "channels_first" on a tf.placeholder with dynamic shapes produces an error, whereas it does not for "channels_last" (see source code below). Cause is usage of reshape operation [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/convolutional.py#L203-L207) as outlined in a similar [issue](https://github.com/tensorflow/tensorflow/issues/15655). Unfortunately, no general solution supporting dynamic shapes was found in aforementioned issue, and shape inference does not allow for >1 dynamic shape.

### Source code / logs
this produces error:
```
import tensorflow as tf
x = tf.placeholder(dtype=tf.float32, shape=[None, 1, None, None, None])
y = tf.layers.conv3d(x, 32, 9, data_format='channels_first')
```

this works fine:
```
import tensorflow as tf
x = tf.placeholder(dtype=tf.float32, shape=[None, None, None, None, 1])
y = tf.layers.conv3d(x, 32, 9, data_format='channels_last')
```

### PR DESCRIPTION
**Pull request by facaiy**
Fix #16834, #15572, #15655, #15696, #16834, #20379, #22771.

Related PR: #21610, #15595

Because #22127 has been merged (thank @yongtang), we can fix the dynamic shape issue easily here.

### CODE REVIEW

File: tensorflow/python/keras/layers/convolutional_test.py [hunk 1]
```diff
@@ -336,6 +336,14 @@ def test_conv3dtranspose_constraints(self):
       self.assertEqual(layer.kernel.constraint, k_constraint)
       self.assertEqual(layer.bias.constraint, b_constraint)
 
+  def test_conv3dtranspose_dynamic_shape(self):
+    with self.session(use_gpu=True):
+      # Won't raise error here.
+      layer = keras.layers.Conv3DTranspose(3, 3, data_format='channels_last')
```
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional_test.py]: This unit test tests nothing. Use `testing_utils.layer_test` instead.

File: tensorflow/python/keras/layers/convolutional.py [hunk 2]
```diff
@@ -199,21 +199,8 @@ def call(self, inputs):
           # nn.bias_add does not accept a 1D input tensor.
           bias = array_ops.reshape(self.bias, (1, self.filters, 1))
           outputs += bias
-        if self.rank == 2:
+        else:
           outputs = nn.bias_add(outputs, self.bias, data_format='NCHW')
-        if self.rank == 3:
-          # As of Mar 2017, direct addition is significantly slower than
```
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional.py]: Has this changed since? Does `bias_add` support 5D inputs in NCDHW mode?

Comment (facaiy) [reply tensorflow/python/keras/layers/convolutional.py]: Yes, `bias_add` has supported 5D input, see PR: https://github.com/tensorflow/tensorflow/pull/22127

File: tensorflow/python/keras/layers/convolutional_test.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/keras/layers/convolutional_test.py]: Fixed. Could you take another look?

**Comment by facaiy** [comment]
@fchollet Hi, could you take another look? Thanks.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
@yongtang Can you please take a look at this PR? Thanks!

**Review by yongtang — APPROVED** [review]
LGTM, Thanks!

**Comment by facaiy** [comment]
Thanks, everybody :-)

**Comment by myron** [comment]
Hi, I don't see these changes in in v1.13-rc1 release, even though it seems to be merged more than a month ago.  are they going to appear in v1.13-rc2 ?

thanks

**Comment by facaiy** [comment]
@myron I think you're right, because r1.3 is cut on 2018-12-13. 

@aselle Can you answer the question about rc2?

**Comment by myron** [comment]
I can this merge is not included in 1.13-rc2 either.

Can you please add this merge to the next rc, or 1.13 final release ??

**Comment by facaiy** [comment]
@myron Because the PR has been merged (and closed). Could you create an issue to request it for 1.13-rc2?

**Comment by myron** [comment]
@facaiy the 1.13-rc2 is already released 6 days ago. 
Do you mean to create an issue to have it in 1.13-rc3 ? (I didn't realize we need a separate issue just to include the already merged updates)

**Comment by facaiy** [comment]
Yes, you have to create an issue for your request if necessary. I'm afraid that the feature will only released in the next version with high probability.  Sorry for the inconvenience.

**Pull request merged**